### PR TITLE
Implement PluginManager (3 of 3)

### PR DESCRIPTION
**Pull request by unode**
Reopened #212 targeting `main` instead of `master`.

### CODE REVIEW

**Comment by unode** [comment]
Some significant changes from #212 - Tests are not yet in sync. More on that in the next push:

* `Plugin.help` and `Plugin.get_help_string` is now `HelpPlugin.help` and `HelpPlugin.get_help_string`.
* The logic to collect help information about plugins is now `PluginManager.get_help`.
* `HelpPlugin` is enabled by default but needs to be explicitly enabled if you specify your own plugins.
* `PluginManager` is now available to `Plugin` instances as `self.manager` initialized via `Plugin.initialize`
* `help_trigger` and `help_trigger_bang` are no more. There's only `settings.RESPOND_CHANNEL_HELP` that defaults to `False` and defines if the bot should respond to `!help` in any channel.
* Arbitrary information can be passed via `listen_to` and `listen_webhook` as `kwargs` and made available via `PluginHelpInfo.annotations`. (e.g. `@listen_to(..., category="webhook", syntax="Command Arg1 Arg2"`)
  * Two attributes are currently recognized by `HelpPlugin.get_help_string`:
    * `category` - grouping functions with the same category label (see screenshot below)
    * `syntax` - providing an alternative to displaying the regex pattern. Instead of `^reply at (.*)$` you can get `reply at TIMESTAMP` or any arbitrary text.
* All help information is now set via function/class docstrings instead of `get_help_string`. The first line of the docstring will be displayed in the help output. The entire docstring is available as `PluginHelpInfo.function_docfull`. Plugin level docstrings are also available, see `plugin_docfull` and `plugin_docheader` 
* `get_help_string` is no longer part of the base ABC `Plugin` class nor `Function` subclasses.
* Docstrings were added to the `ExamplePlugin` where missing.

The current help output with plugins `HelpPlugin`, `ExamplePlugin` and `WebHookExample` looks like:
screenshot_2021-05-31_03-27-06_647421429

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+          - "--in-place"
+          - "--wrap-summaries"
+          - "88"
+          - "--wrap-descriptions"
+          - "88"
+  - repo: https://github.com/pycqa/doc8
+    # sphinx rst style checker
+    rev: 0.9.0a1
+    hooks:
+      - id: doc8
```
Comment (jneeven) [inline .pre-commit-config.yaml:56]: I'm fine with this (although I think the formatted lines in the docs are wrapped *way* too aggressively), but could you do this in a different PR? It's not related to the PluginManager and the many doc changes make reviewing a bit annoying...

Comment (unode) [reply .pre-commit-config.yaml:56]: I found it a bit annoying as well. Will pull it out into another PR.

Comment (unode) [reply .pre-commit-config.yaml:56]: Now in #228 . I'll rebase this PR once the other is merged.

**Comment by unode** [comment]
I think this is ready for a final of review.

**Comment by unode** [comment]
This PR will also close #165

**Comment by attzonko** [comment]
Again @unode thanks for putting in all this work. I have a quick request, can you elaborate on what you mean by the bullet below and what exactly does it mean for users who use their own plugins: 
>  `HelpPlugin` is enabled by default but needs to be explicitly enabled if you specify your own plugins.

**Comment by unode** [comment]
> Again @unode thanks for putting in all this work. I have a quick request, can you elaborate on what you mean by the bullet below and what exactly does it mean for users who use their own plugins:
> 
> > `HelpPlugin` is enabled by default but needs to be explicitly enabled if you specify your own plugins.

I mean this https://github.com/attzonko/mmpy_bot/blob/a584bf242673f7355b25b32f6d0afe0d33b76fb7/mmpy_bot/bot.py#L37

If you run the bot without adding any custom plugins, it will be included among the default plugins.
If do add custom plugins (the most likely case) you should include it, or a subclass of it, on the list of plugins to enable.

Initially I thought of having it always enabled (i.e. `plugins + [HelpPlugin()]`) but that would remove the ability to define or customize your own help plugin.

File: mmpy_bot/plugins/base.py [hunk 2]
```diff
@@ -88,21 +71,185 @@ async def call_function(
...
+            for attribute in dir(plugin):
+                attribute = getattr(plugin, attribute)
+                if not isinstance(attribute, Function):
+                    continue
+
+                # Register this function and any potential siblings
+                for function in [attribute] + attribute.siblings:
+                    # Plugin message/webhook handlers can be decorated multiple times
+                    # resulting in multiple siblings that do not have .plugin defined
+                    # or where the relationship with the parent plugin is incorrect
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:193]: What do you mean by this? Has this changed?

File: mmpy_bot/plugins/base.py [hunk 3]
```diff
@@ -88,21 +71,185 @@ async def call_function(
...
+                        self.webhook_listeners[function.matcher].append(function)
+                    else:
+                        raise TypeError(
+                            f"{plugin.__class__.__name__} has a function of unsupported"
+                            f" type {type(function)}."
+                        )
+
+    def _split_docstring(self, doc):
+        """Split docstring into first line (header) and full body."""
+        return (doc.split("\n", 1)[0], doc) if doc is not None else ("", "")
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:207]: Since this is not class-specific and does not depend on any `self` variables, can you move this out of the class?

File: mmpy_bot/plugins/base.py [hunk 4]
```diff
@@ -88,21 +71,185 @@ async def call_function(
...
+
+    def _generate_plugin_help(
+        self,
+        plug_help: List[PluginHelpInfo],
+        help_type: str,
+        items: ItemsView[re.Pattern, List[Function]],
+    ):
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:214]: Same here, could you move this out of the class?

Also, why is `plug_help` modified in place and can't it be created and returned here?

File: mmpy_bot/plugins/base.py [hunk 5]
```diff
@@ -88,21 +71,185 @@ async def call_function(
...
+
+    def _split_docstring(self, doc):
+        """Split docstring into first line (header) and full body."""
+        return (doc.split("\n", 1)[0], doc) if doc is not None else ("", "")
+
+    def _generate_plugin_help(
+        self,
+        plug_help: List[PluginHelpInfo],
+        help_type: str,
+        items: ItemsView[re.Pattern, List[Function]],
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:213]: Just a personal preference, but I think it's cleaner to just loop over the original dict here rather than specifically requiring an `ItemsView`. My default assumption is that a function does not modify its arguments in place.

File: mmpy_bot/plugins/base.py [hunk 6]
```diff
@@ -88,21 +71,185 @@ async def call_function(
...
+                if help_type == "message":
+                    direct = function.direct_only
+                    mention = function.needs_mention
+                elif help_type == "webhook":
+                    direct = mention = False
+                else:
+                    raise NotImplementedError(f"Unknown help type: '{help_type}'")
+
+                plug_help.append(
+                    PluginHelpInfo(
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:233]: The class is called `PluginHelpInfo`, but it looks like this is actually a `FunctionInfo`?

File: mmpy_bot/plugins/base.py [hunk 7]
```diff
@@ -88,21 +71,185 @@ async def call_function(
...
+                        is_click=function.is_click_function,
+                        annotations=function.annotations,
+                    )
+                )
+
+    def get_help(self) -> List[PluginHelpInfo]:
+        response: List[PluginHelpInfo] = []
+
+        self._generate_plugin_help(response, "message", self.message_listeners.items())
+        self._generate_plugin_help(response, "webhook", self.webhook_listeners.items())
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:253]: I think list concatenation is way cleaner here, see my comment above

File: mmpy_bot/plugins/example.py [hunk 8]
```diff
@@ -14,12 +14,14 @@
 class ExamplePlugin(Plugin):
     """Default plugin with examples of bot functionality and usage."""
 
-    @listen_to("^admin$", direct_only=True, allowed_users=["admin", "root"])
+    @listen_to(
+        "^admin$", direct_only=True, allowed_users=["admin", "root"], category="admin"
```
Comment (jneeven) [inline mmpy_bot/plugins/example.py:18]: I like the `category` addition!

File: mmpy_bot/plugins/example.py [hunk 9]
```diff
@@ -28,14 +30,20 @@ async def busy_reply(self, message: Message):
             f"Number of busy worker threads: {busy}",
         )
 
-    @listen_to("hello_click", needs_mention=True)
+    @listen_to(
+        "hello_click",
+        needs_mention=True,
+        category="click",
+        syntax="hello_click --keyword-arg --flag",
```
Comment (jneeven) [inline mmpy_bot/plugins/example.py:37]: What does this `syntax` argument do? We should be able to obtain this information from `click` rather than having to hardcode it?

File: mmpy_bot/plugins/example.py [hunk 10]
```diff
@@ -110,7 +119,13 @@ async def ping_reply(self, message: Message):
         """Pong."""
         self.driver.reply_to(message, "pong")
 
-    @listen_to("^reply at (.*)$", re.IGNORECASE, needs_mention=True)
+    @listen_to(
+        "^reply at (.*)$",
+        re.IGNORECASE,
+        needs_mention=True,
+        category="schedule",
+        syntax="reply at TIMESTAMP",
```
Comment (jneeven) [inline mmpy_bot/plugins/example.py:127]: I'd prefer storing this kind of information in the docstring rather than an explicit `syntax` argument here; it feels unnatural.

File: tests/unit_tests/plugin_manager_test.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    re.compile("pattern"): expand_func_names(FakePlugin.my_function),
+    re.compile("direct_pattern"): expand_func_names(FakePlugin.direct_function),
+    re.compile("another_async_pattern"): expand_func_names(
+        FakePlugin.my_async_function
+    ),
+    re.compile("async_pattern"): expand_func_names(FakePlugin.my_async_function),
+    re.compile("hi_custom"): expand_func_names(FakePlugin.hi_custom),
+    # Click commands construct a regex pattern from the listen_to pattern
+    re.compile("^click_command (.*)?"): expand_func_names(FakePlugin.click_commmand),
+}
```
Comment (jneeven) [inline tests/unit_tests/plugin_manager_test.py:71]: Why are these redefined? Can't we just take the arguments passed to `listen_to`?

File: mmpy_bot/function.py [hunk 12]
```diff
@@ -5,22 +5,27 @@
...
 log = logging.getLogger("mmpy.function")
 
 
 class Function(ABC):
     def __init__(
         self,
-        function: Callable,
+        function: Union[Callable, Function],
         matcher: re.Pattern,
+        **annotations,
```
Comment (jneeven) [inline mmpy_bot/function.py:28]: Can you rename this `metadata`? `annotations` suggests that these might be the type annotations of the wrapped function 😅

File: mmpy_bot/function.py [hunk 13]
```diff
@@ -80,6 +80,8 @@ def __init__(
         else:
             self.allowed_channels = [channel.lower() for channel in allowed_channels]
 
+        self.docstring = self.function.__doc__ or ""
```
Comment (jneeven) [inline mmpy_bot/function.py:83]: Why can this not be defined on the parent class?

File: mmpy_bot/function.py [hunk 14]
```diff
@@ -216,15 +188,20 @@ def wrapped_func(func):
...
             matcher=pattern,
             direct_only=direct_only,
             needs_mention=needs_mention,
             allowed_users=allowed_users,
             allowed_channels=allowed_channels,
+            **annotations,
         )
 
+        # Preserve docstring
+        new_func.__doc__ = func.__doc__
```
Comment (jneeven) [inline mmpy_bot/function.py:202]: Why is this necessary? We can definitely explicitly store the `__doc__` of the wrapped function somewhere, but I'd prefer not to modify the `__doc__` of the wrapping `Function` since `__doc__` is normally class-specific, not instance-specific.

File: mmpy_bot/plugins/base.py [hunk 15]
```diff
@@ -88,21 +71,185 @@ async def call_function(
...
+                rec.annotations.get("category", ""),  # No categories first
+                rec.help_type,
+                rec.pattern.lstrip("^[(-"),
+            )
+
+        string = "### The following functions have been registered:\n\n"
+        string += "###### `(*)` require the use of `@botname`, "
+        string += "`(+)` can only be used in direct message\n"
+        old_category = None
+        for h in sorted(self.manager.get_help(), key=custom_sort):
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:108]: If we're just using `self.manager.get_help()` here, why can't we move all this logic there? Is `PluginManager.get_help` used anywhere else?

File: tests/unit_tests/plugin_manager_test.py [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+        ]
+
+
+def create_message(
+    text="hello",
+    mentions=["qmw86q7qsjriura9jos75i4why"],
+    channel_type="O",
+    sender_name="betty",
+):
+    return Message(
```
Comment (jneeven) [inline tests/unit_tests/plugin_manager_test.py:121]: Why can't this be imported from the event handler test? If you prefer, we could move this to some `utils` file as well.

**Review by jneeven — COMMENTED** [review]
Nice, thanks for all the work 🙌🏼! I think having a separate `PluginManager` is great, I just have a bunch of questions and suggestions here and there. It's a huge PR, so there are loads of comments 🙃  In the future it'd be nicer to split a big one like this into maybe ~4 PRs, that makes it much easier to see what's going on. I haven't look at all the details yet, so will probably have to do another round of reviewing later.

File: mmpy_bot/plugins/base.py (hunk 15 above)
Comment (unode) [reply mmpy_bot/plugins/base.py:108]: Initially I tried implementing help on the `PluginManager` class but due to how listeners are registered the manager itself would have to behave as a `Plugin` to respond to "help". I tried going with this exception but didn't like the result so I eventually converged in `HelpPlugin`.

The separation I tried to do here is:
1. `PluginManager.get_help()` collects the help and creates `PluginHelpInfo` instances for them.
2. `HelpPlugin.get_help_string()` implement transforms the above into a string. This method can be subclassed or re-implemented in other plugins by users. 

I couldn't come up with any useful way to allow partial reuse of the `HelpPlugin.get_help_string()` code. Could maybe pull out parts of the text but it felt like too much guessing for potential use-cases.

I also think it makes more sense to have this in the plugin as an example of how to go about implementing it.

File: tests/unit_tests/plugin_manager_test.py (hunk 16 above)
Comment (unode) [reply tests/unit_tests/plugin_manager_test.py:121]: Accidental leftover it's not actually used. Removed.

File: mmpy_bot/function.py (hunk 13 above)
Comment (unode) [reply mmpy_bot/function.py]: Done

File: mmpy_bot/function.py (hunk 12 above)
Comment (unode) [reply mmpy_bot/function.py]: Done

File: mmpy_bot/bot.py [hunk 17]
```diff
@@ -95,7 +107,7 @@ def run(self):
             if self.settings.WEBHOOK_HOST_ENABLED:
                 self.driver.threadpool.start_webhook_server_thread(self.webhook_server)
 
-            for plugin in self.plugins:
+            for plugin in self.manager:
                 plugin.on_start()
```
[the remark this answers is not in the document]
Comment (unode) [reply mmpy_bot/bot.py]: Move iteration to .start() and .stop() and removed the iterable.

File: mmpy_bot/bot.py [hunk 18]
```diff
@@ -54,23 +69,20 @@ def __init__(
...
+        self.manager.initialize_manager(self.driver, self.settings)
         self.event_handler = EventHandler(
-            self.driver, settings=self.settings, plugins=self.plugins
+            self.driver, settings=self.settings, manager=self.manager
         )
         self.webhook_server = None
 
-        if self.settings.WEBHOOK_HOST_ENABLED:
-            self._initialize_webhook_server()
+        self._initialize_webhook_server()
```
[the remark this answers is not in the document]
Comment (unode) [reply mmpy_bot/bot.py]: Reverted. I wasn't too happy with that conditional in the init body but you point makes a lot of sense.

File: mmpy_bot/bot.py [hunk 19]
```diff
@@ -22,10 +28,19 @@ class Bot:
...
         self,
         settings: Optional[Settings] = None,
-        plugins: Optional[Sequence[Plugin]] = None,
+        plugins: Optional[Union[List[Plugin], PluginManager]] = None,
     ):
+        self.manager: PluginManager
+
         if plugins is None:
-            plugins = [ExamplePlugin(), WebHookExample()]
+            self.manager = PluginManager(
```
[the remark this answers is not in the document]
Comment (unode) [reply mmpy_bot/bot.py]: Done. Makes things a bit more verbose now but I guess that's ok.

File: mmpy_bot/bot.py [hunk 20]
```diff
@@ -22,10 +28,19 @@ class Bot:
     def __init__(
         self,
         settings: Optional[Settings] = None,
-        plugins: Optional[Sequence[Plugin]] = None,
+        plugins: Optional[Union[List[Plugin], PluginManager]] = None,
     ):
+        self.manager: PluginManager
+
```
[the remark this answers is not in the document]
Comment (unode) [reply mmpy_bot/bot.py]: Done

File: mmpy_bot/function.py (hunk 14 above)
Comment (unode) [reply mmpy_bot/function.py:200]: This one was tricky. We now have `PluginHelpInfo.function_doc[header|full]` as well as `PluginHelpInfo.plugin_doc[header|full]` which contain the method and class level `Plugin` docstring, respectively.
With the way decorated functions are transformed into `MessageFunction` or `WebHookFunction` and there can be multiple decorators (multiple listen patterns, click decorators, ...), this was the simplest way to propagate the docstring without having to do some kind of conditional code using `isinstance()`.

The test `tests/unit_tests/plugin_manager_test.py::TestPluginManager::test_get_help` verifies if the docstring is propagated correctly. Commenting these lines breaks it.

**Review by unode — COMMENTED** [review]
> Nice, thanks for all the work 🙌🏼! I think having a separate `PluginManager` is great, I just have a bunch of questions and suggestions here and there. It's a huge PR, so there are loads of comments upside_down_face In the future it'd be nicer to split a big one like this into maybe ~4 PRs, that makes it much easier to see what's going on. I haven't look at all the details yet, so will probably have to do another round of reviewing later.

Thanks a lot Jelmer. Your detailed review is super appreciated.
Indeed this grew more than expected so, take your time. I also took mine :sweat_smile:  .

I tried to get different PRs at some point but wasn't clear how to go about dependencies between PRs. I can pull some things out if you think it would make your life easier.

**Comment by unode** [comment]
Pulled #233 out of this one. That's a pretty self contained feature.

**Comment by unode** [comment]
After trying to customize a little the help info I noticed the current approach actually forces you to modify `PluginManager.get_help` if you want to filter out some functions and `HelpPlugin.get_help_string` if you want to customize the format of the help.

To simplify the use-case of filtering listeners, I now introduced a `HelpPlugin.get_help` method. It currently only calls to `PluginManager.get_help` but provides a location for users to override what is included without having to redefine `HelpPlugin.get_help_string` entirely.
See also https://github.com/attzonko/mmpy_bot/pull/226/files#diff-6cc66ec79b3b320572cf875b2e985afab611933a46d625218fc030c3a000131dR188 as an example of such use-case.

File: mmpy_bot/function.py [hunk 21]
```diff
@@ -92,12 +92,9 @@ def __init__(
                 info_name=self.matcher.pattern.strip("^").split(" (.*)?")[0],
             ) as ctx:
                 # Get click help string and do some extra formatting
-                self.docstring = self.function.get_help(ctx).replace(
-                    "\n", f"\n{spaces(8)}"
-                )
+                self.docstring += f"\n\n{self.function.get_help(ctx)}"
```
Comment (jneeven) [inline mmpy_bot/function.py:95]: I think this is what causes those duplicate docstrings you showed earlier in Discord. `click` already includes the docstring in `get_help`, so this shouldn't be `+=`.

```suggestion
                self.docstring = f"\n\n{self.function.get_help(ctx)}"
```

File: mmpy_bot/function.py [hunk 22]
```diff
@@ -30,26 +35,21 @@ def __init__(
             self.siblings.append(function)
             function = function.function
 
-        self.function = function
+        self.function: Union[Callable, Function] = function
```
Comment (jneeven) [inline mmpy_bot/function.py:38]:
```suggestion
        self.function = function
```
The type is already annotated just a few lines above. Also, given the function above, I believe `function` is guaranteed to just be a `Callable` here (if it was still a Function, the loop would have just triggered another time).

File: mmpy_bot/function.py (hunk 14 above)
Comment (jneeven) [reply mmpy_bot/function.py:200]: I still don't understand this. The base `Function` class has `self.docstring = self.function.__doc__ or ""`, where `self.function` is always the absolute innermost function (i.e. if you pass it a Function, it will keep unwrapping it until it reaches the callable that this Function wraps). Why then would we need to change `self.__doc__` as well, and can't we just use `self.docstring`?

File: mmpy_bot/plugins/base.py [hunk 23]
```diff
@@ -88,21 +70,213 @@ async def call_function(
...
-            for function in functions:
-                string += f"- {function.get_help_string()}"
-            string += "----\n"
-        if len(self.webhook_listeners) > 0:
-            string += "### Registered webhooks:\n"
-            for functions in self.webhook_listeners.values():
-                for function in functions:
-                    string += f"- {function.get_help_string()}"
+
+class HelpPlugin(Plugin):
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:74]: Can you move this Plugin to its own file? `ExamplePlugin` has its own file too, and this one is getting quite large.

File: mmpy_bot/plugins/base.py [hunk 24]
```diff
@@ -88,21 +70,213 @@ async def call_function(
...
+        driver: Driver,
+        plugin_manager: PluginManager,
+        settings: Settings,
+    ):
+        super().initialize(driver, plugin_manager, settings)
+
+        if self.settings.RESPOND_CHANNEL_HELP:
+            self.help = listen_to("^!help$")(self.help)
+
+    def get_help_string(self, message: Message) -> str:
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:102]: Given the very similar name to `get_help` below, can you add a docstring here?

File: mmpy_bot/plugins/base.py [hunk 25]
```diff
@@ -88,21 +70,213 @@ async def call_function(
...
+
+            # Register listeners for any listener functions in the plugin
+            for attribute in dir(plugin):
+                attribute = getattr(plugin, attribute)
+                if not isinstance(attribute, Function):
+                    continue
+
+                # Register this function and any potential siblings
+                for function in [attribute] + attribute.siblings:
+                    # Plugin message/webhook handlers can be decorated multiple times
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:254]: Why have you moved this functionality out of the `Plugin` class? I think accessing some `Plugin` attributes in a for loop here (generated by e.g. `Plugin.initialize()`) is a lot cleaner than this triple-nested for loop...

File: mmpy_bot/plugins/base.py [hunk 26]
```diff
@@ -88,21 +70,213 @@ async def call_function(
...
+
+
+@dataclass
+class FunctionInfo:
+    help_type: str
+    location: str
+    function: Function
+    pattern: str
+    plugin_docheader: str
+    plugin_docfull: str
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:164]: Storing this information for every function in the plugin is redundant, and I believe these values are actually not used anywhere? Then I'd prefer removing them.

File: mmpy_bot/plugins/base.py [hunk 27]
```diff
@@ -88,21 +70,213 @@ async def call_function(
...
+                    function=function,
+                    pattern=matcher.pattern,
+                    plugin_docheader=plug_head,
+                    plugin_docfull=plug_full,
+                    function_docheader=func_head,
+                    function_docfull=func_full,
+                    direct=direct,
+                    mention=mention,
+                    is_click=function.is_click_function,
+                    metadata=function.metadata,
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:213]: Since all of these attributes are accessible on the `Function` itself, I don't see the added value of this `FunctionInfo` class. Instead of storing this all in that class here, you might as well change the `HelpPlugin` to directly obtain this from the `Function`?

File: mmpy_bot/plugins/example.py [hunk 28]
```diff
@@ -110,7 +119,13 @@ async def ping_reply(self, message: Message):
...
         self.driver.reply_to(message, "pong")
 
-    @listen_to("^reply at (.*)$", re.IGNORECASE, needs_mention=True)
+    @listen_to(
+        "^reply at (.*)$",
+        re.IGNORECASE,
+        needs_mention=True,
+        category="schedule",
+        human_description="reply at TIMESTAMP",
+    )
```
Comment (jneeven) [inline mmpy_bot/plugins/example.py:128]: I understand that you want this info for the `HelpPlugin`, but I don't like how verbose this has made the `ExamplePlugin`. The `ExamplePlugin` serves as a showcase of how easily you can create various kinds of behaviour, and the massive decorators with optional additional information take away from that purpose a bit... Are you okay with removing this for most of these functions and instead just relying on the documentation to explain the help behaviour?

File: tests/unit_tests/event_handler_test.py [hunk 29]
```diff
@@ -127,10 +103,14 @@ def test_handle_event(self, handle_post):
     @mock.patch("mmpy_bot.driver.Driver.username", new="my_username")
     def test_handle_post(self):
         # Create an initialized plugin so its listeners are registered
+        plugin = ExamplePlugin()
         driver = Driver()
```
Comment (jneeven) [inline tests/unit_tests/event_handler_test.py:107]:
```suggestion
        driver = Driver()
```

File: tests/unit_tests/event_handler_test.py [hunk 30]
```diff
@@ -127,10 +103,14 @@ def test_handle_event(self, handle_post):
...
-        plugin = ExamplePlugin().initialize(driver)
+        plugin_manager = PluginManager([plugin])
```
Comment (jneeven) [inline tests/unit_tests/event_handler_test.py:108]:
```suggestion
        plugin_manager = PluginManager([ExamplePlugin()])
```

File: tests/unit_tests/event_handler_test.py [hunk 31]
```diff
@@ -153,10 +133,14 @@ async def mock_call_function(function, message, groups):
 
     def test_handle_webhook(self):
         # Create an initialized plugin so its listeners are registered
+        plugin = WebHookExample()
         driver = Driver()
-        plugin = WebHookExample().initialize(driver, Settings())
+        plugin_manager = PluginManager([plugin])
```
Comment (jneeven) [inline tests/unit_tests/event_handler_test.py:138]:
```suggestion
        plugin_manager = PluginManager([WebHookExample()])
```

File: tests/unit_tests/event_handler_test.py [hunk 32]
```diff
@@ -153,10 +133,14 @@ async def mock_call_function(function, message, groups):
 
     def test_handle_webhook(self):
         # Create an initialized plugin so its listeners are registered
+        plugin = WebHookExample()
         driver = Driver()
```
Comment (jneeven) [inline tests/unit_tests/event_handler_test.py:137]:
```suggestion
        driver = Driver()
```

File: mmpy_bot/plugins/base.py [hunk 33]
```diff
@@ -25,39 +27,19 @@ class Plugin(ABC):
...
-        return self
+    def initialize(
+        self,
+        driver: Driver,
+        plugin_manager: PluginManager,
+        settings: Settings,
+    ):
+        self.driver = driver
+        self.plugin_manager = plugin_manager
+        self.settings = settings
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:42]:
```suggestion
        self.settings = settings
```

The `Plugin` does not actually need the manager for anything, only the `HelpPlugin` does, and therefore that one can just override this function (which it already does anyway)

File: mmpy_bot/plugins/base.py [hunk 34]
```diff
@@ -88,21 +70,213 @@ async def call_function(
...
+        self.direct_help: bool = direct_help
+
+    def initialize(
+        self,
+        driver: Driver,
+        plugin_manager: PluginManager,
+        settings: Settings,
+    ):
+        super().initialize(driver, plugin_manager, settings)
+
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:98]:
```suggestion
self.function_info = self.plugin_manager.get_help()
```

File: mmpy_bot/plugins/base.py [hunk 35]
```diff
@@ -88,21 +70,213 @@ async def call_function(
...
         return string
 
+    def get_help(self, message: Message):
+        """Obtain Help info from PluginManager.
+
+        Override this method if you need to customize which listeners will be included
+        in the help.
+        """
+        return self.plugin_manager.get_help()
+
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:148]:
```suggestion
```

Can be deleted now that `initialize` takes care of this.

File: mmpy_bot/plugins/base.py [hunk 36]
```diff
@@ -88,21 +70,213 @@ async def call_function(
...
+    def get_help_string(self, message: Message) -> str:
+        def custom_sort(rec):
+            return (
+                rec.metadata.get("category", ""),  # No categories first
+                rec.help_type,
+                rec.pattern.lstrip("^[(-"),
+            )
+
+        help_function_info = sorted(self.get_help(message), key=custom_sort)
```
Comment (jneeven) [inline mmpy_bot/plugins/base.py:110]:
```suggestion
        help_function_info = sorted(self.function_info, key=custom_sort)
```

File: tests/unit_tests/plugin_manager_test.py (hunk 11 above)
Comment (jneeven) [reply tests/unit_tests/plugin_manager_test.py:71]: Did you accidentally link the wrong file? I don't see the relevance of that line here. I think the relevant test was [this one](https://github.com/attzonko/mmpy_bot/blob/main/tests/unit_tests/event_handler_test.py#L59), and it did not require duplicating the listeners like this

**Review by jneeven — COMMENTED** [review]
Round two 😅 These reviews are taking me a long time, so I would be grateful if you could either make PRs into this branch and/or not force-push to it, so I can look only at the latest changes

**Comment by unode** [comment]
> Round two sweat_smile These reviews are taking me a long time, so I would be grateful if you could either make PRs into this branch and/or not force-push to it, so I can look only at the latest changes

Sorry, I will stop pulling things in/out until the review is complete.

File: mmpy_bot/function.py (hunk 21 above)
Comment (unode) [reply mmpy_bot/function.py:95]: I tried `=` before but then we lose all of the help from `click`.
I'm not happy with either of the solutions but I think we'll need to address `click` UI/UX in another PR.
I agree that `click` is convenient for argument parsing but the help information is (IMO) too verbose for a chat-bot.

File: tests/unit_tests/plugin_manager_test.py (hunk 11 above)
Comment (unode) [reply tests/unit_tests/plugin_manager_test.py:71]: Sorry I read your comment in a different light. I was justifying the why regexp and click listeners are treated differently with a link to the code where click listeners are transformed into a 'catch-all' regexp.

As to why I re-defined the list of expected listeners... The original test was comparing information stored in `Plugin` instances with that on the event handler. Since there's now only one place where listeners are collected, the only way I could think to test this was by either repeating the logic already used by the `PluginManager` or hardcoding the expected result. The first option didn't make sense to me so I went with the second.

File: mmpy_bot/plugins/base.py (hunk 24 above)
Comment (unode) [reply mmpy_bot/plugins/base.py]: Done

File: mmpy_bot/function.py (hunk 14 above)
Comment (unode) [reply mmpy_bot/function.py:200]: There might be some confusion on my part. When I tried to capture this in a test, the test fails unless I included the line above.

Can you perhaps have a look at the test `tests/unit_tests/plugin_manager_test.py::TestPluginManager::test_get_help` and confirm that the test and its assumptions are correct?

File: mmpy_bot/function.py (hunk 21 above)
Comment (jneeven) [reply mmpy_bot/function.py:95]: > I tried = before but then we lose all of the help from click.

What do you mean? The "help from click" is what is returned by `self.function.get_help(ctx)` right?

File: mmpy_bot/plugins/base.py (hunk 27 above)
Comment (jneeven) [reply mmpy_bot/plugins/base.py]: Okay, fair enough :+1:

File: tests/unit_tests/plugin_manager_test.py (hunk 11 above)
Comment (jneeven) [reply tests/unit_tests/plugin_manager_test.py:71]: I see. I still believe that the logic of finding the listeners on any given Plugin should be in a function on the Plugin itself, even if it will only be called by the PluginManager. That also removes the need for any code duplication here.

File: tests/unit_tests/plugin_manager_test.py [hunk 37]
```diff
@@ -0,0 +1,169 @@
...
+
+        assert len(hook_listeners) == len(self.plugin_manager.webhook_listeners)
+        for pattern, listeners in self.plugin_manager.webhook_listeners.items():
+            for listener in listeners:
+                assert pattern in hook_listeners
+                assert listener in hook_listeners[pattern]
+
+    def test_get_help(self):
+        # Prior to initialization there is no help
+        assert self.plugin_manager.get_help() == []
```
Comment (jneeven) [inline tests/unit_tests/plugin_manager_test.py:138]: This seems to rely on `self.plugin_manager` not having been initialized, but it's initialized first thing in `test_init`?

File: mmpy_bot/function.py (hunk 14 above)
Comment (jneeven) [reply mmpy_bot/function.py:200]: I can't  judge what's going on just from looking at the code, so will have to try running it later

**Review by jneeven — COMMENTED** [review]
Your recent changes look good to me and I've resolved some comments. Some comments I left earlier are still relevant though, could you have another look at those?

I think we're nearing convergence here, so I propose the following:
- Let's try to first resolve all comments open in this PR. I won't add any new ones, but we can continue discussing the ones that already have some discussion going on.
- Once everything is resolved, I'll need to have a fresh look at all the changes here, and since this is an enormous PR and there have been months in between each time I looked at it, it would be best to split this into several easily reviewable PRs so it becomes very clear what changes are being made in what context, and maybe @attzonko  can then have a look at it as well if he wants. I think there are easily several PRs worth of changes here, such as the introduction of metadata, the introduction of the FunctionInfo, etc. If the interdependence between all these things is an issue, feel free to open one initial PR of very limited scope, and then make the others PR merge into that one rather than directly into `main`. We can then wait with merging that central PR until all the individual changes have been reviewed and merged into it without needing to look at 17 files at a time :sweat_smile:

File: mmpy_bot/function.py (hunk 21 above)
Comment (unode) [reply mmpy_bot/function.py:95]: Yes to `self.function.get_help(ctx)`. If memory serves the issue was functions decorated with both `@listen_to` and `@click...`. The inner click decorated is processed and `function.get_help(ctx)` is called but as the outer decorator (`listen_to`) is processed later, this was overwriting what was done before, hence the `+=`.

File: tests/unit_tests/plugin_manager_test.py (hunk 11 above)
Comment (unode) [reply tests/unit_tests/plugin_manager_test.py:71]: That's doable but with it we need to do some extra work later when building the help page. i.e. we will need to traverse all the plugins every time, sort/group, etc... instead of just calling to the `PluginManager`.
I remember I started with this approach but ran into some situations either I duplicated information both in `Plugin` and `PluginManager` or I was constantly looping over all plugins.
If you think listener initialization should stay with `Plugin` we might need to bring the duplication back to avoid a possible performance hit with looping for almost every event.

File: tests/unit_tests/plugin_manager_test.py (hunk 37 above)
Comment (unode) [reply tests/unit_tests/plugin_manager_test.py:138]: Doesn't pytest run the two tests independently? i.e.
```
t = TestPluginManager()
t.setup_method()
t.test_get_help()

t = TestPluginManager()
t.setup_method()
t.test_get_init()
```

**Comment by unode** [comment]
Thanks @jneeven!

I'll give it another try at pulling out different PRs. Will likely go with your suggestion of making the PRs against the previous.

**Comment by aconitumnapellus** [comment]
Im eagerly awaiting this PR! Currently have a .help command to show what commands are available and what they do

**Comment by unode** [comment]
See #323 and #324 for an attempt to split this PR.
The current history will be rebased on the other two PRs once merged.
